File: app/src/main/java/com/vocaby/app/database/VocabyDatabase.java
package com.vocaby.app.database;

import android.content.Context;
import android.content.SharedPreferences;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.room.Database;
import androidx.room.Room;
import androidx.room.RoomDatabase;
import androidx.sqlite.db.SupportSQLiteDatabase;

import com.vocaby.app.database.dao.VocabyDao;
import com.vocaby.app.database.entity.Definition;
import com.vocaby.app.database.entity.User;
import com.vocaby.app.database.entity.UserSaves;
import com.vocaby.app.database.entity.Word;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers;
import io.reactivex.rxjava3.schedulers.Schedulers;

@Database(entities = {Word.class, User.class, Definition.class, UserSaves.class}, version = 2, exportSchema = false)
public abstract class VocabyDatabase extends RoomDatabase {
    public abstract VocabyDao vocabyDao();

    private static volatile VocabyDatabase INSTANCE;
    public static VocabyDatabase getDatabase(final Context context) {
        if (INSTANCE == null) {
            synchronized (VocabyDatabase.class) {
                if (INSTANCE == null) {
                    INSTANCE = Room.databaseBuilder(context.getApplicationContext(),
                            VocabyDatabase.class, "vocaby_database")
                            .createFromAsset("databases/vocabydevdb.db")
                            // .fallbackToDestructiveMigration()
                            .build();
                }
            }
        }

        return INSTANCE;
    }
}

File: app/src/main/java/com/vocaby/application/core/data/VocabyDatabase.kt
package com.vocaby.application.core.data

import android.content.Context
import androidx.room.Database
import androidx.room.Room
import androidx.room.RoomDatabase
import androidx.room.TypeConverters
import androidx.sqlite.db.SupportSQLiteDatabase
import com.vocaby.application.feature_dictionary.data.local.DictionaryDao
import com.vocaby.application.feature_dictionary.data.local.entity.Definition
import com.vocaby.application.feature_dictionary.data.local.entity.Type
import com.vocaby.application.feature_dictionary.data.local.entity.Word
import com.vocaby.application.feature_dictionary_custom.data.local.CustomDictionaryDao
import com.vocaby.application.feature_dictionary_custom.data.local.entity.CustomDefinition
import com.vocaby.application.feature_dictionary_custom.data.local.entity.CustomEntry
import com.vocaby.application.feature_dictionary_custom.data.local.entity.CustomEntryGroup
import com.vocaby.application.feature_profile.data.local.UserDao
import com.vocaby.application.feature_profile.data.local.entity.CustomDictionaryViewCount
import com.vocaby.application.feature_profile.data.local.entity.DictionaryViewCount
import com.vocaby.application.feature_profile.data.local.entity.User
import com.vocaby.application.feature_save.data.local.SaveDao
import com.vocaby.application.feature_save.data.local.entity.SaveCollection
import com.vocaby.application.feature_save.data.local.entity.SaveCollectionItem
import com.vocaby.application.feature_save.data.local.entity.UserSave
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch

@Database(entities = [User::class, UserSave::class, SaveCollection::class, SaveCollectionItem::class,
    Type::class, Word::class, Definition::class, CustomEntry::class, CustomDefinition::class,
    CustomEntryGroup::class, DictionaryViewCount::class, CustomDictionaryViewCount::class],
    version = 1,
    exportSchema = true,
)
@TypeConverters(Converters::class)
abstract class VocabyDatabase : RoomDatabase() {
    abstract val dictionaryDao: DictionaryDao
    abstract val customDictionaryDao: CustomDictionaryDao
    abstract val userDao: UserDao
    abstract val saveDao: SaveDao

    companion object {
        fun getDatabase(context: Context, availableTypes: List<Type>): VocabyDatabase {
            return Room.databaseBuilder(
                context,
                VocabyDatabase::class.java,
                "database"
            ).createFromAsset("databases/vocabydb.db")
            .setJournalMode(JournalMode.AUTOMATIC)
            .addCallback(object: Callback() {
                override fun onCreate(db: SupportSQLiteDatabase) {
                    val applicationScope = CoroutineScope(SupervisorJob())
                    applicationScope.launch {
                        getDatabase(context, availableTypes).customDictionaryDao.insertTypes(*availableTypes.toTypedArray())
                    }
                }
            })
            .build()

        }
    }
}